server/block/smelter.go

```diff
@@ -0,0 +1,170 @@
+package block
+
+import (
+	"github.com/df-mc/dragonfly/server/block/cube"
+	"github.com/df-mc/dragonfly/server/item"
+	"github.com/df-mc/dragonfly/server/item/inventory"
+	"github.com/df-mc/dragonfly/server/world"
+	"math"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+// smelter is a struct that may be embedded by blocks that can smelt blocks and items, such as blast furnaces, furnaces,
+// and smokers.
+type smelter struct {
+	mu sync.Mutex
+
+	viewers   map[ContainerViewer]struct{}
+	inventory *inventory.Inventory
+
+	remainingDuration time.Duration
+	cookDuration      time.Duration
+	maxDuration       time.Duration
+	experience        int
+}
+
+// newSmelter initializes a new smelter with the given remaining, maximum, and cook durations and XP, and returns it.
+func newSmelter() *smelter {
+	s := &smelter{viewers: make(map[ContainerViewer]struct{})}
+	s.inventory = inventory.New(3, func(slot int, item item.Stack) {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		for viewer := range s.viewers {
+			viewer.ViewSlotChange(slot, item)
+		}
+	})
+	return s
+}
+
+// SetExperience sets the collected experience of the smelter to the given value.
+func (s *smelter) SetExperience(xp int) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.experience = xp
+}
+
+// SetDurations sets the remaining, maximum, and cook durations of the smelter to the given values.
+func (s *smelter) SetDurations(remaining, max, cook time.Duration) {
```

Comment:
This method I think is also only used within this package. Would make it unexported too.

Reply:
Unexported.